subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/worker/device/Device.kt

```diff
@@ -35,9 +36,16 @@ interface Device {
 
     /**
      * @return `to` path
+     *
+     * todo deprecate in favor of [pullDirRecursively]
      */
     fun pull(from: Path, to: Path): Result<File>
 
+    /**
+     * @return `hostDir` path
+     */
+    fun pullDirRecursively(deviceDir: Path, hostDir: Path, validator: PullValidator): Result<File>
```

Comment:
We force clients to pass `validator` on each invocation.
But it seems like details of copying.
Can't we hide from API?

Reply:
every pull operation can have it's own validator; we can't

Reply:
> every pull operation can have it's own validator

I see this in the implementation but don't understand restrictions or use cases.

Can you please bring an example of when a client needs to use different validators? As I assume, on the same device.

Reply:
One instance of `AdbDevice` can be used for multiple pull operations through single run; for now it's only pulling report artifacts, maybe we will want to pull something else with other validation rules.

Reply:
Agree, didn't think about different directories. Validation is directory\operation specific.